### PR TITLE
Add support for CQFM Group Id extension (composite measure report building)

### PR DESCRIPTION
**Pull request by sarahmcdougall**
# Summary
For composite measure report building, filters the component results to those corresponding to groups defined by the CQFM Group Id extension (if applicable).

## New behavior
Resources:
* [Composites page in the Quality Measure IG](https://build.fhir.org/ig/HL7/cqf-measures/composite-measures.html) 
* [CQFM Composite Measure Resource Profile](https://build.fhir.org/ig/HL7/cqf-measures/StructureDefinition-composite-measure-cqfm.html)
* [CQFM Group Id Extension Definition](https://build.fhir.org/ig/HL7/cqf-measures/StructureDefinition-cqfm-groupId.html)

From the [Multiple Populations](https://build.fhir.org/ig/HL7/cqf-measures/composite-measures.html#multiple-populations) section of the spec: "Component measures used in a composite SHOULD contain a single population group. For component measures that contain multiple population groups, the composite measure SHALL specify the specific group to be used in the composite using the cqfm-groupId extension."

Behavior does not change for calculating detailed results. Behavior *does* change for composite measure report building. When calculating the measure score, the component results will first be filtered so that the population results used for calculating the measure score are those that correspond to the specified groups. In other words, rather than considering all the component results across all the defined groups for measure score calculation, only one component result for a single group will be considered for measure score calculation. The logic can be broken down as follows:
* If the component only contains one population group, that group is used to get the info necessary for measure report building
* If the component contains multiple population groups and the CQFM Group Id extension is specified for the component, then the specified group will be used to get the info necessary for measure report building
* If the component contains multiple populations groups and the CQFM Group Id extension is *not* defined for the component, an error is thrown

## Code changes
* `MeasureBundleHelpers` - helper functions that extract the group Id `valueString` from the defined CQFM Group Id extension and filter the component results according to the defined group Ids on the extension
* `CompositeReportBuilder` - calls helper functions from `MeasureBundleHelpers` to extract the group Ids and filter the component results before looping through them for measure score calculation
* New unit tests for extracting the group Id from the extension and for filtering the component results according to the defined groups on the extension

# Testing guidance
Since no (known) composite measures exist in the wild, the best way to test functionality will be to run unit tests and integration tests. That being said, testing guidance is as follows:
* Run unit tests —> tests that the Group Id extension can successfully be extracted and that the component results can be appropriately filtered
* Run integration tests —> the integration tests do not test measure report building, but running them ensures that the detailed results are not filtered in any way (we only want to do filtering for report building)
* To test via the CLI, refer to attached `.zip`. The `.zip` contains a directory of patients and two different measure bundles:
    * `composite-mb-no-extension.json` - does not define the group Id extension. Component 1 has two defined groups. Should throw an error since a specific group is not specified with the extension.
    * `composite-mb-with-groupId.json` - defines the group Id extension for Component 1. Component 1 has two defined groups. Should successfully build a composite measure report using the defined group.
    * Test out edge cases by tweaking these measure bundles. For example, change the group Id extension in `composite-mb-with-groupId.json` to specify a non-existent group. This should throw an error.
    * Test across the different composite scoring types. This can be done by changing the value of `compositeScoring.coding.code` on the composite measure in the measure bundle. At a minimum, test one of the scoring types that is checked in `addPatientResults` (all or nothing, opportunity, or linear) and test the `weighted` scoring type since that gets calculated in `addWeightedResults`.

[composite.zip](https://github.com/projecttacoma/fqm-execution/files/11546938/composite.zip)

### CODE REVIEW

**Review by elsaperelli — CHANGES_REQUESTED** [review]
I was trying some different test cases for this with @mgramigna and found an issue with the test case where there are two related artifacts of the same component (`resource` is the same) but the extensions of them are different groups. An example of this can be found at the bottom of the composite measures IG:

image

Notice the `resource` values of the two related artifacts are the same. Therefore, in `CompositeReportBuilder.ts` when the `componentGroupIds` object gets constructed, the value is overwritten. 

If you run measure reports on the `composite-mb-with-groupId.json` measure bundle that you provide on just the `pt-proc-only.json` patient, the measure score is 0.5. If you add a related artifact for for component-1-group-2, then the measure score should be different (specifically, 2/3), but it is currently the same.

Attached below is the edited measure bundle I used to test this!

[composite-mb-with-2-groups.json.zip](https://github.com/projecttacoma/fqm-execution/files/11558210/composite-mb-with-2-groups.json.zip)

File: src/helpers/MeasureBundleHelpers.ts [hunk 1]
```diff
@@ -107,6 +107,61 @@ export function extractCompositeMeasure(measureBundle: fhir4.Bundle): fhir4.Meas
...
+ * @param componentGroupIds mapping of group ids defined by the CQFM Group Id extension on the composite
+ * @param componentResults array of component results from detailed results
+ * @returns component results, filtered by desired group Ids
+ */
+export function filterComponentResults(
+  componentGroupIds: Record<string, string>,
+  componentResults?: ComponentResults[]
+): ComponentResults[] {
+  const filteredComponentResults: ComponentResults[] = [];
+  componentResults?.forEach(cr => {
```
Comment (lmd59) [inline src/helpers/MeasureBundleHelpers.ts:136]: I think that this might be more straightforward using `Array.filter()` rather than pushing things onto `filteredComponentResults using `forEach`...
Probably would not be too big a change. Something like:
```suggestion
  const filteredComponentResults = componentResults?.filter(cr => {
```
... and then instead of pushing cr onto filteredComponentResults, you would return true in those cases.

**Comment by sarahmcdougall** [comment]
> I was trying some different test cases for this with @mgramigna and found an issue with the test case where there are two related artifacts of the same component (`resource` is the same) but the extensions of them are different groups. An example of this can be found at the bottom of the composite measures IG:

Ah, great catch. This example falls under [part 2(c) here](https://build.fhir.org/ig/HL7/cqf-measures/composite-measures.html#component-quality-measures): "Multiple groups within the same measure may be referenced as different components of the same composite." I forgot this was possible because I keep thinking of measure and component as being synonymous, which isn't always the case since a measure can be referenced within multiple components. I'll update the storage method for mapping components to group Id

**Review by elsaperelli — CHANGES_REQUESTED** [review]
I think I found an issue with the weighted scoring calculation with groupId extension (not your fault, these maybe should not have been done at the same time lol). If we have the test case that I mentioned before where we have three related artifacts where two of the them are the same component but different groupIds, then in the weighted scoring calculation only two components are counted which messed up the denominator. Attached is the example I am referring to. If component-1-group-1 has weight of 0.3 and component-1-group-2 has weight of 0.2 and component-2 does not have a weight (so weight is 1)...then I believe the calculation should be 0.3(2/3) + 0.2(2/3) +1(2/3) / 3...I am not sure how the numerator is currently calculated (it is 0.799), but I think the denominator is only counting the 2 components because it is 2 but it should be 3.

My guess is we have to handle numer and denom counts better for weighted scoring if there are groupId extensions.

[groupId-weighted.json.zip](https://github.com/projecttacoma/fqm-execution/files/11578563/groupId-weighted.json.zip)

**Review by elsaperelli — CHANGES_REQUESTED** [review]
Woo looks all good to me after lots of testing!!

I have one overall suggestion for how to store the component canonicals, their groupIds (if they have them) and their weights (also if they have them). I discussed this a bit with @mgramigna and I think it makes a little more sense but most importantly eliminates the need for `JSON.stringify` :)

So rather than have `weightedComponents` and `componentGroupids`, thoughts on combining them in a Record of type `Record<string, Record<string, number>>`? This looks pretty daunting but I played around with it and I think it actually works. The following code is just one idea for how building this object from the relatedArtifacts may look:

image

In the end, the `this.components` object would look something like this:

image


Then, in the `addWeightedResults` function, you can check the type of `this.components[ra.resource]`, if it is a number, then just use that weight, but if it is `Record<string, number>`, then you could access the weight for a specific group in that Record. 

Sorry that was a lot, let me know if you want to discuss!

**Review by elsaperelli — CHANGES_REQUESTED** [review]
Only small comments, I promise!!! Sorry this one has dragged on for so long- almost there! Testing all works :)

File: src/calculation/CompositeReportBuilder.ts [hunk 2]
```diff
@@ -103,29 +118,91 @@ export class CompositeReportBuilder<T extends PopulationGroupResult> extends Abs
...
-            componentResult.populationResults?.find(pr => pr.populationType === PopulationType.NUMER)?.result === true
-          ) {
-            componentPopulationResults[componentResult.componentCanonical].numerator++;
+      const componentResults = filterComponentResults(this.components, result.componentResults);
+      componentResults.forEach(componentResult => {
+        if (componentResult.componentCanonical) {
+          // component info will either consist of mapping of groups to weights, or a single weight
+          const componentInfo = this.components[componentResult.componentCanonical];
+          // if only one weight is specified, the component corresponds to a single group
+          const hasSingleGroup = typeof componentInfo === 'number';
```
[the remark this answers is not in the document]
Comment (sarahmcdougall) [reply src/calculation/CompositeReportBuilder.ts:135]: Good idea - I think `noGroupExt` is more representative of what the boolean represents

File: src/calculation/CompositeReportBuilder.ts [hunk 3]
```diff
@@ -103,29 +118,91 @@ export class CompositeReportBuilder<T extends PopulationGroupResult> extends Abs
...
+          if (hasSingleGroup) {
+            if (!(componentResult.componentCanonical in componentPopulationResults)) {
```
[the remark this answers is not in the document]
Comment (sarahmcdougall) [reply src/calculation/CompositeReportBuilder.ts:137]: hmm what would the error be that we would throw in this case?
Here, if the group id extension is not defined, then we just want to check if the component canonical's weight has already been captured in the `componentPopulationResults` record or not. If not, we add the canonical as a key and map it to the weight. Otherwise, we do not do anything since we already have the weight info for the component, and we can continue with incrementing the numer and denom. 

The reason why we could come across the canonical more than once is because we are looping over the Execution results, meaning we have results for this canonical for each patient. In the screenshot below, you can see that when looping over each patient, we run across the `measure-ComponentTwo` canonical. We only add it as key to the record the first time that we encounter it.
Screenshot 2023-06-05 at 8 48 45 AM

Theoretically, the way that calculation is set up, I would assume that each patient has component results for all the components (unless they are filtered out beforehand based on the groupid ext, which in this case is not present). I guess, rather than throwing an error, we could have a separate loop that initializes the `componentPopulationResults` record so that we do not need to loop through all the records unnecessarily? I kept the initial structure that was used for the weighted scoring and tailored it to the new groupid functionality, which may no longer be the best approach

[the remark this answers is not in the document]
Comment (sarahmcdougall) [reply src/calculation/CompositeReportBuilder.ts:137]: In the above screenshot, I added `console.log`s to lines 129, 138, and 139 in `CompositeReportBuilder.ts` if you want to replicate

**Review by lmd59 — APPROVED** [review]
Yay! Great PR updates!
